File: web/src/extractor.js
/**
 * Port of zotprep/extractor.py — raw reference string in, ParsedRef out.
 *
 * The Python original opens by noting that the whole accuracy strategy rests on
 * this file: everything downstream compares candidates against these extracted
 * fields, so a bad parse here becomes a bad match later. That makes it the
 * second-most parity-sensitive file after the string metric, and the one where
 * Python and JavaScript regex semantics differ most.
 *
 * The differences that actually bite, and how each is handled:
 *
 *   \b and \w      Python's are Unicode-aware on `str`; JavaScript's are
 *                  ASCII-only. `\bPMID\b` in Python does NOT match inside
 *                  "PMIDé" (é is a word character) but the JS equivalent does.
 *                  Every word boundary here is therefore written explicitly as
 *                  a `(?<![\p{L}\p{N}_])` / `(?![\p{L}\p{N}_])` pair, and `\w`
 *                  as `[\p{L}\p{N}_]`, with the `u` flag.
 *
 *   str.replace    Python's replaces *every* occurrence; JavaScript's replaces
 *                  only the first when given a string. The identifier-stripping
 *                  step relies on the Python behaviour, so it uses split/join.
 *
 *   $              Python's `$` also matches before a trailing newline;
 *                  JavaScript's does not. Inputs here are whitespace-collapsed
 *                  by nfkc() first, so no trailing newline ever reaches these
 *                  patterns.
 *
 *   isupper/islower  Unicode-aware in Python. Written as \p{Lu} / \p{Ll}.
 *
 *   str.strip(chars)  removes any of several characters; String.trim() does not.
 *
 * Verified field-by-field against the Python original by
 * tools/parity/extractor_parity.py.
 */
import { ParsedRef } from './models.js';
import { rstripChars, stripChars } from './utils.js';

// Unicode-aware word-boundary assertions, matching Python's \b on str.
const NWB = '(?<![\\p{L}\\p{N}_])';
const NWA = '(?![\\p{L}\\p{N}_])';
const W = '[\\p{L}\\p{N}_]';

// --- identifier patterns -----------------------------------------------------
// DOI: trailing punctuation is stripped afterwards, since refs end in "."
const DOI_RE = new RegExp(`${NWB}10\\.\\d{4,9}/[^\\s"<>,;]+`, 'iu');
const DOI_TRAILING = /[.,;:)\]}>'"]+$/;
const PMID_RE = new RegExp(`${NWB}PMID:?\\s*(\\d{4,9})${NWA}`, 'iu');
const PMID_RE_G = new RegExp(PMID_RE.source, 'giu');
const PMCID_RE = new RegExp(`${NWB}(PMC\\d{4,9})${NWA}`, 'iu');
const PMCID_RE_G = new RegExp(PMCID_RE.source, 'giu');

// --- structural patterns -----------------------------------------------------
const LEAD_NUM = /^\s*[[(]?\s*(\d{1,3})\s*[.)\]]\s+/;
const BRACKET_NOTE = /\[[^\]]*\]/g;
const YEAR_RE = new RegExp(`${NWB}(19|20)\\d{2}${NWA}`, 'gu');

// "2017;390(10111):2437-60"  /  "2012;10:1"  /  "2004;S2:11-44"
const LOCATOR_SRC =
  `${NWB}(?<year>(?:19|20)\\d{2})\\s*;\\s*` +
  '(?<vol>[A-Za-z]?[\\dA-Za-z]{0,8}?)\\s*' +
  '(?:\\(\\s*(?<issue>[^)]{1,20})\\s*\\))?\\s*' +
  ':\\s*(?<fp>[A-Za-z]?\\d+)(?:\\s*-\\s*(?<lp>[A-Za-z]?\\d+))?';
const LOCATOR_RE = new RegExp(LOCATOR_SRC, 'u');

// Book imprint tail. Both separators before the year occur in the wild:
//   "Amsterdam: North-Holland; 1967"      (Vancouver)
//   "Amsterdam: North-Holland, 1967"      (Lancet house style)
//   "Amsterdam: Elsevier, 2013: 1113-36"  (chapter, with page range)
const BOOK_TAIL_RE = new RegExp(
  '^(?<place>[^:;]{2,40}):\\s*(?<publisher>[^;,]{2,60})[;,]\\s*' +
  '(?<year>(?:19|20)\\d{2})' +
  '(?:\\s*:\\s*(?<fp>[A-Za-z]?\\d+)(?:\\s*-\\s*(?<lp>[A-Za-z]?\\d+))?)?\\.?$',
  'u',
);
const EDITION_RE = /^(?<ed>\d+(?:st|nd|rd|th)|rev(?:ised)?|\d+)\s*edn?\.?$/iu;
// trailing ", 4th edn" attached to a book title rather than standing alone
const INLINE_EDITION_RE = /,\s*(?<ed>\d+(?:st|nd|rd|th)|rev(?:ised)?)\s*edn?\.?\s*$/iu;
const INLINE_EDITION_RE_G = new RegExp(INLINE_EDITION_RE.source, 'giu');
const CHAPTER_IN_RE = /^In:\s*(?<editors>.+?),?\s*eds?\.?$/iu;

const CORPORATE_HINT = new RegExp(
  `${NWB}(?:collaborators?|collaboration|group|consortium|committee|initiative|` +
  'investigators|network|study team|working party|organization|organisation|' +
  `who|unicef|world health)${NWA}`,
  'iu',
);

// A segment is an author list if it looks like "Surname AB, Surname CD" —
// i.e. mostly "word + 1-4 capital initials" tokens.
const AUTHOR_TOKEN = new RegExp(
  `^(?<sur>(?:[A-Z]${W}*(?:\\s+[a-z]{2,3})?(?:[-\\s][A-Z]${W}*)*))` +
  '\\s+(?<init>(?:[A-Z]\\.?){1,4})$',
  'u',
);
// Python's `[\w'’-]` — \w plus apostrophes and hyphen. Rebuilt here because the
// class above interpolates W; kept identical in content and order.
const AUTHOR_TOKEN_FULL = new RegExp(
  `^(?<sur>(?:[A-Z][\\p{L}\\p{N}_'’-]*(?:\\s+[a-z]{2,3})?(?:[-\\s][A-Z][\\p{L}\\p{N}_'’-]*)*))` +
  '\\s+(?<init>(?:[A-Z]\\.?){1,4})$',
  'u',
);
const SENTENCE_SPLIT = /(?<=[.?!])\s+/;
const INITIALS_ONLY = /^(?:[A-Z]\.?){1,4}$/u;
const ET_AL = /,?\s*et al\.?$/iu;

/** Python's `str.split()` with no argument. */
function words(s) {
  const t = String(s ?? '').trim();
  return t ? t.split(/\s+/) : [];
}

/** Normalize unicode and unify the dash/quote zoo found in pasted refs. */
export function nfkc(s) {
  let t = String(s ?? '').normalize('NFKC');
  for (const bad of '‐‑‒–—―−') t = t.split(bad).join('-');
  for (const [bad, good] of [['‘', "'"], ['’', "'"], ['“', '"'], ['”', '"']]) {
    t = t.split(bad).join(good);
  }
  return t.replace(/\s+/g, ' ').trim();
}

/** Vancouver abbreviates end pages: 2437-60 means 2437-2460, e1339-51 -> e1351. */
export function expandPage(first, last) {
  if (!first || !last) return last ?? null;
  const preF = /^([A-Za-z]*)(\d+)$/.exec(first);
  const preL = /^([A-Za-z]*)(\d+)$/.exec(last);
  if (!(preF && preL)) return last;
  const fdig = preF[2];
  let ldig = preL[2];
  if (ldig.length < fdig.length) ldig = fdig.slice(0, fdig.length - ldig.length) + ldig;
  return (preL[1] || preF[1]) + ldig;
}

/**
 * Split a reference into sentence-ish segments.
 *
 * Vancouver puts a period after the last author's initials and after "et al.",
 * so a plain sentence split lands exactly on the author/title/journal
 * boundaries. Mid-list authors are separated by commas, not periods, so they
 * survive the split intact.
 */
function splitSegments(text) {
  return text.split(SENTENCE_SPLIT).map((p) => p.trim()).filter(Boolean);
}

function looksLikeAuthors(seg) {
  let s = rstripChars(seg, '.');
  s = s.replace(ET_AL, '');
  const chunks = s.split(',').map((c) => c.trim()).filter(Boolean);
  if (!chunks.length) return false;
  const hits = chunks.filter((c) => AUTHOR_TOKEN_FULL.test(c)).length;
  return hits >= Math.max(1, Math.floor(chunks.length / 2));
}

/** Return [personal surnames, corporate name or null]. */
function parseAuthors(seg) {
  let s = rstripChars(seg.trim(), '.');
  s = s.replace(ET_AL, '');
  if (CORPORATE_HINT.test(s) && !looksLikeAuthors(s + '.')) return [[], s];
  const surnames = [];
  for (const rawChunk of s.split(',')) {
    const chunk = rawChunk.trim();
    if (!chunk) continue;
    const m = AUTHOR_TOKEN_FULL.exec(chunk);
    if (m) {
      surnames.push(m.groups.sur.trim());
    } else {
      // "Surname, Given" (APA) or an un-initialled name
      const parts = words(chunk);
      const word = parts.length ? parts[0] : '';
      if (/^\p{Lu}/u.test(word.slice(0, 1)) && word.length > 1 && !INITIALS_ONLY.test(word)) {
        surnames.push(word);
      }
    }
  }
  if (!surnames.length && CORPORATE_HINT.test(s)) return [[], s];
  return [surnames, null];
}

export function parseReference(n, raw) {
  const ref = new ParsedRef(n, String(raw).trim());
  let text = nfkc(raw);
  text = text.replace(LEAD_NUM, '');

  // --- identifiers: pull, then remove so they don't pollute segmentation ---
  let m = DOI_RE.exec(text);
  if (m) {
    ref.doi = m[0].replace(DOI_TRAILING, '').toLowerCase();
    // Python's str.replace removes every occurrence, not just the first
    text = text.split(m[0]).join(' ');
  }
  m = PMID_RE.exec(text);
  if (m) {
    ref.pmid = m[1];
    text = text.replace(PMID_RE_G, ' ');
  }
  m = PMCID_RE.exec(text);
  if (m) {
    ref.pmcid = m[1].toUpperCase();
    text = text.replace(PMCID_RE_G, ' ');
  }

  // editorial annotations like "[Zotero: complete vol/pages/DOI]" are noise
  text = text.replace(BRACKET_NOTE, ' ');
  text = text.replace(new RegExp(`${NWB}(?:doi|available from|accessed)${NWA}[:.]?\\s*`, 'giu'), ' ');
  text = text.replace(/https?:\/\/\S+/g, ' ');
  text = text.replace(/\s+/g, ' ').trim();

  // --- locator (journal;volume(issue):pages) -------------------------------
  const loc = LOCATOR_RE.exec(text);
  if (loc) {
    const g = loc.groups;
    ref.year = parseInt(g.year, 10);
    ref.volume = (g.vol || '').trim() || null;
    ref.issue = (g.issue || '').trim() || null;
    ref.first_page = g.fp ?? null;
    ref.last_page = expandPage(g.fp, g.lp);
  }

  const segments = splitSegments(text);
  if (!segments.length) return ref;

  // --- author segment ------------------------------------------------------
  const [authors, corporate] = parseAuthors(segments[0]);
  ref.authors = authors;
  ref.corporate = corporate;
  const body = segments.slice(1);

  // --- book chapter: "In: <editors>, eds. <book title>[, Nth edn]. <imprint>"
  for (let i = 0; i < body.length; i++) {
    if (CHAPTER_IN_RE.test(rstripChars(body[i], '.') + '.')) {
      ref.is_book = true;
      ref.is_chapter = true;
      ref.title = rstripChars(body.slice(0, i).join(' '), '.').trim();
      const rest = body.slice(i + 1);
      if (rest.length) {
        let bt = rstripChars(rest[0], '.').trim();
        const em = INLINE_EDITION_RE.exec(bt);
        if (em) {
          ref.edition = em.groups.ed;
          bt = bt.replace(INLINE_EDITION_RE_G, '').trim();
        }
        ref.book_title = bt;
      }
      for (const seg2 of rest.slice(1)) {
        const bm = BOOK_TAIL_RE.exec(seg2.trim());
        if (bm) {
          ref.place = bm.groups.place.trim();
          ref.publisher = bm.groups.publisher.trim();
          ref.year = parseInt(bm.groups.year, 10);
          ref.first_page = bm.groups.fp ?? null;
          ref.last_page = expandPage(bm.groups.fp, bm.groups.lp);
          break;
        }
      }
      ref.title = stripChars(ref.title.replace(/\s+/g, ' '), ' .,;');
      return ref;
    }
  }

  // --- edition marker ------------------------------------------------------
  for (let i = 0; i < body.length; i++) {
    const em = EDITION_RE.exec(rstripChars(body[i], '.'));
    if (em) {
      ref.edition = em.groups.ed;
      body.splice(i, 1);
      break;
    }
  }

  // --- locate the tail: either a locator segment or a book imprint ---------
  let tailIdx = null;
  for (let i = 0; i < body.length; i++) {
    if (LOCATOR_RE.test(body[i])) { tailIdx = i; break; }
    const bm = BOOK_TAIL_RE.exec(body[i]);
    if (bm) {
      ref.is_book = true;
      ref.place = bm.groups.place.trim();
      ref.publisher = bm.groups.publisher.trim();
      ref.year = ref.year ?? parseInt(bm.groups.year, 10);
      if (bm.groups.fp) {
        ref.first_page = ref.first_page ?? bm.groups.fp;
        ref.last_page = ref.last_page ?? expandPage(bm.groups.fp, bm.groups.lp);
      }
      tailIdx = i;
      break;
    }
  }

  if (tailIdx === null) {
    // No locator and no imprint: fall back to the last segment that is just a
    // year (e.g. "Lancet. 2024.") — then journal is the segment before it.
    for (let i = body.length - 1; i >= 0; i--) {
      if (/^(?:19|20)\d{2}\.?$/u.test(body[i].trim())) {
        tailIdx = i;
        ref.year = ref.year ?? parseInt(rstripChars(body[i].trim(), '.'), 10);
        break;
      }
    }
  }

  if (tailIdx === null) {
    // last resort: title is everything, year from anywhere in the string
    ref.title = rstripChars(body.join(' '), '.').trim();
  } else if (ref.is_book) {
    ref.title = rstripChars(body.slice(0, tailIdx).join(' '), '.').trim();
  } else {
    // segment immediately before the tail is the journal, unless the tail
    // segment itself carries the journal ("Lancet 2017;390(1):1-2")
    const jrnIdx = tailIdx - 1;
    const tailSeg = body[tailIdx];
    const locHit = LOCATOR_RE.exec(tailSeg);
    const preLocator = locHit ? stripChars(tailSeg.slice(0, locHit.index), ' ,.;') : '';
    if (preLocator && !/^(?:19|20)\d{2}$/u.test(preLocator)) {
      ref.journal = preLocator;
      ref.title = rstripChars(body.slice(0, tailIdx).join(' '), '.').trim();
    } else if (jrnIdx >= 0) {
      ref.journal = rstripChars(body[jrnIdx], '.').trim();
      ref.title = rstripChars(body.slice(0, jrnIdx).join(' '), '.').trim();
    } else {
      ref.title = rstripChars(body.slice(0, tailIdx).join(' '), '.').trim();
    }
  }

  if (ref.year === null) {
    const years = text.match(YEAR_RE);
    if (years && years.length) ref.year = parseInt(years[years.length - 1], 10);
  }

  // a book with no imprint but an edition marker is still a book
  if (!ref.is_book && ref.publisher === null && ref.volume === null && ref.edition) {
    ref.is_book = true;
  }

  ref.title = stripChars(ref.title.replace(/\s+/g, ' '), ' .,;');
  ref.journal = stripChars(ref.journal.replace(/\s+/g, ' '), ' .,;');
  return ref;
}

export function parseAll(refs) {
  const out = new Map();
  for (const [n, txt] of refs instanceof Map ? refs : Object.entries(refs)) {
    out.set(Number(n), parseReference(Number(n), txt));
  }
  return out;
}

File: zotprep/extractor.py
"""Turn a raw reference string into a ParsedRef.

The whole accuracy strategy rests on this file. Everything downstream compares
candidates against these extracted fields, so a bad parse here becomes a bad
match later. Deliberately conservative: when a field cannot be extracted with
confidence it is left None rather than guessed, because the scorer treats a
missing signal as neutral but a wrong signal as evidence.

Handles Vancouver (primary), plus APA/AMA-ish and book forms:

    Barro RJ, Sala-i-Martin X. Convergence. J Polit Econ. 1992;100(2):223-51.
    GBD 2021 Collaborators. Global incidence ... Lancet. 2024.
    Theil H. Economics and Information Theory. Amsterdam: North-Holland; 1967.
    Cowell FA. Measuring Inequality. 3rd ed. Oxford: Oxford University Press; 2011.
"""
from __future__ import annotations

import re
import unicodedata

from .models import ParsedRef

# --- identifier patterns -----------------------------------------------------
# DOI: trailing punctuation is stripped afterwards, since refs end in "."
DOI_RE = re.compile(r"\b10\.\d{4,9}/[^\s\"<>,;]+", re.I)
DOI_TRAILING = re.compile(r"[.,;:)\]}>'\"]+$")
PMID_RE = re.compile(r"\bPMID:?\s*(\d{4,9})\b", re.I)
PMCID_RE = re.compile(r"\b(PMC\d{4,9})\b", re.I)

# --- structural patterns -----------------------------------------------------
LEAD_NUM = re.compile(r"^\s*[\[(]?\s*(\d{1,3})\s*[.)\]]\s+")
BRACKET_NOTE = re.compile(r"\[[^\]]*\]")
YEAR_RE = re.compile(r"\b(19|20)\d{2}\b")

# "2017;390(10111):2437-60"  /  "2012;10:1"  /  "2004;S2:11-44"
#
# Two shapes that are ordinary in medical Vancouver and that a tighter pattern
# misses entirely — and missing the locator does not degrade the match, it
# removes it: journal, volume and first page all come back empty, so the accept
# gate has nothing left to agree on and the reference is rejected.
#
#   "2011;63 Suppl 11:S240-52"   a supplement, with a space inside the volume
#   "2016;10(9):UC05-7"          a first page with two letters, not one
#   "2018;100-B(8):991-1001"     a lettered volume, as the Bone & Joint
#                                Journal numbers them — and Crossref stores it
#                                that way too, so the suffix is kept
#
# The supplement is captured so it cannot be mistaken for part of the volume,
# and then dropped: indexes store the base volume ("63"), which is what the
# volume comparison needs to see.
LOCATOR_RE = re.compile(
    r"\b(?P<year>(?:19|20)\d{2})\s*;\s*"
    r"(?P<vol>[A-Za-z]?[\dA-Za-z]{0,8}?(?:-[A-Za-z]{1,2})?)"
    r"(?:\s*(?P<suppl>[Ss]uppl(?:ement)?\.?\s*[\dA-Za-z]{0,4}))?\s*"
    r"(?:\(\s*(?P<issue>[^)]{1,20})\s*\))?\s*"
    r":\s*(?P<fp>[A-Za-z]{0,3}\d+)(?:\s*-\s*(?P<lp>[A-Za-z]{0,3}\d+))?"
)
# Book imprint tail. Both separators before the year occur in the wild:
#   "Amsterdam: North-Holland; 1967"   (Vancouver)
#   "Amsterdam: North-Holland, 1967"   (Lancet house style)
#   "Amsterdam: Elsevier, 2013: 1113-36"  (chapter, with page range)
BOOK_TAIL_RE = re.compile(
    r"^(?P<place>[^:;]{2,40}):\s*(?P<publisher>[^;,]{2,60})[;,]\s*"
    r"(?P<year>(?:19|20)\d{2})"
    r"(?:\s*:\s*(?P<fp>[A-Za-z]?\d+)(?:\s*-\s*(?P<lp>[A-Za-z]?\d+))?)?\.?$"
)
EDITION_RE = re.compile(r"^(?P<ed>\d+(?:st|nd|rd|th)|rev(?:ised)?|\d+)\s*edn?\.?$", re.I)
# trailing ", 4th edn" attached to a book title rather than standing alone
INLINE_EDITION_RE = re.compile(r",\s*(?P<ed>\d+(?:st|nd|rd|th)|rev(?:ised)?)\s*edn?\.?\s*$", re.I)
CHAPTER_IN_RE = re.compile(r"^In:\s*(?P<editors>.+?),?\s*eds?\.?$", re.I)

CORPORATE_HINT = re.compile(
    r"\b(collaborators?|collaboration|group|consortium|committee|initiative|"
    r"investigators|network|study team|working party|organization|organisation|"
    r"who|unicef|world health)\b",
    re.I,
)

# A segment is an author list if it looks like "Surname AB, Surname CD" —
# i.e. mostly "word + 1-4 capital initials" tokens.
AUTHOR_TOKEN = re.compile(
    r"^(?P<sur>(?:[A-Z][\w'’-]*(?:\s+[a-z]{2,3})?(?:[-\s][A-Z][\w'’-]*)*))"
    r"\s+(?P<init>(?:[A-Z]\.?){1,4})$"
)
SENTENCE_SPLIT = re.compile(r"(?<=[.?!])\s+")


def _nfkc(s: str) -> str:
    """Normalize unicode and unify the dash/quote zoo found in pasted refs."""
    s = unicodedata.normalize("NFKC", s)
    for bad in "‐‑‒–—―−":
        s = s.replace(bad, "-")
    for bad, good in (("‘", "'"), ("’", "'"), ("“", '"'), ("”", '"')):
        s = s.replace(bad, good)
    return re.sub(r"\s+", " ", s).strip()


def expand_page(first: str | None, last: str | None) -> str | None:
    """Vancouver abbreviates end pages: 2437-60 means 2437-2460, e1339-51 -> e1351."""
    if not first or not last:
        return last
    pre_f = re.match(r"^([A-Za-z]*)(\d+)$", first)
    pre_l = re.match(r"^([A-Za-z]*)(\d+)$", last)
    if not (pre_f and pre_l):
        return last
    fdig, ldig = pre_f.group(2), pre_l.group(2)
    if len(ldig) < len(fdig):
        ldig = fdig[: len(fdig) - len(ldig)] + ldig
    return (pre_l.group(1) or pre_f.group(1)) + ldig


def _split_segments(text: str) -> list[str]:
    """Split a reference into sentence-ish segments.

    Vancouver puts a period after the last author's initials and after "et al.",
    so a plain sentence split lands exactly on the author/title/journal
    boundaries. Mid-list authors are separated by commas, not periods, so they
    survive the split intact.
    """
    return [p.strip() for p in SENTENCE_SPLIT.split(text) if p.strip()]


def _looks_like_authors(seg: str) -> bool:
    seg = seg.rstrip(".")
    seg = re.sub(r",?\s*et al\.?$", "", seg, flags=re.I)
    chunks = [c.strip() for c in seg.split(",") if c.strip()]
    if not chunks:
        return False
    hits = sum(1 for c in chunks if AUTHOR_TOKEN.match(c))
    return hits >= max(1, len(chunks) // 2)


def _parse_authors(seg: str) -> tuple[list[str], str | None]:
    """Return (personal surnames, corporate name or None)."""
    seg = seg.strip().rstrip(".")
    seg = re.sub(r",?\s*et al\.?$", "", seg, flags=re.I)
    if CORPORATE_HINT.search(seg) and not _looks_like_authors(seg + "."):
        return [], seg
    surnames: list[str] = []
    for chunk in seg.split(","):
        chunk = chunk.strip()
        if not chunk:
            continue
        m = AUTHOR_TOKEN.match(chunk)
        if m:
            surnames.append(m.group("sur").strip())
        else:
            # "Surname, Given" (APA) or an un-initialled name
            word = chunk.split()[0] if chunk.split() else ""
            if word[:1].isupper() and len(word) > 1 and not re.fullmatch(r"(?:[A-Z]\.?){1,4}", word):
                surnames.append(word)
    if not surnames and CORPORATE_HINT.search(seg):
        return [], seg
    return surnames, None


def parse_reference(n: int, raw: str) -> ParsedRef:
    ref = ParsedRef(n=n, raw=raw.strip())
    text = _nfkc(raw)
    text = LEAD_NUM.sub("", text)

    # --- identifiers: pull, then remove so they don't pollute segmentation ---
    m = DOI_RE.search(text)
    if m:
        ref.doi = DOI_TRAILING.sub("", m.group(0)).lower()
        text = text.replace(m.group(0), " ")
    m = PMID_RE.search(text)
    if m:
        ref.pmid = m.group(1)
        text = PMID_RE.sub(" ", text)
    m = PMCID_RE.search(text)
    if m:
        ref.pmcid = m.group(1).upper()
        text = PMCID_RE.sub(" ", text)

    # editorial annotations like "[Zotero: complete vol/pages/DOI]" are noise
    text = BRACKET_NOTE.sub(" ", text)
    text = re.sub(r"\b(?:doi|available from|accessed)\b[:.]?\s*", " ", text, flags=re.I)
    text = re.sub(r"https?://\S+", " ", text)
    text = re.sub(r"\s+", " ", text).strip()

    # --- locator (journal;volume(issue):pages) -------------------------------
    loc = LOCATOR_RE.search(text)
    if loc:
        ref.year = int(loc.group("year"))
        ref.volume = (loc.group("vol") or "").strip() or None
        ref.issue = (loc.group("issue") or "").strip() or None
        ref.first_page = loc.group("fp")
        ref.last_page = expand_page(loc.group("fp"), loc.group("lp"))

    segments = _split_segments(text)
    if not segments:
        return ref

    # --- author segment ------------------------------------------------------
    ref.authors, ref.corporate = _parse_authors(segments[0])
    body = segments[1:]

    # --- book chapter: "In: <editors>, eds. <book title>[, Nth edn]. <imprint>"
    for i, seg in enumerate(body):
        if CHAPTER_IN_RE.match(seg.rstrip(".") + "."):
            ref.is_book = ref.is_chapter = True
            ref.title = " ".join(body[:i]).rstrip(".").strip()
            rest = body[i + 1:]
            if rest:
                bt = rest[0].rstrip(".").strip()
                em = INLINE_EDITION_RE.search(bt)
                if em:
                    ref.edition = em.group("ed")
                    bt = INLINE_EDITION_RE.sub("", bt).strip()
                ref.book_title = bt
            for seg2 in rest[1:]:
                bm = BOOK_TAIL_RE.match(seg2.strip())
                if bm:
                    ref.place = bm.group("place").strip()
                    ref.publisher = bm.group("publisher").strip()
                    ref.year = int(bm.group("year"))
                    ref.first_page = bm.group("fp")
                    ref.last_page = expand_page(bm.group("fp"), bm.group("lp"))
                    break
            ref.title = re.sub(r"\s+", " ", ref.title).strip(" .,;")
            return ref

    # --- edition marker ------------------------------------------------------
    for i, seg in enumerate(list(body)):
        em = EDITION_RE.match(seg.rstrip("."))
        if em:
            ref.edition = em.group("ed")
            body.pop(i)
            break

    # --- locate the tail: either a locator segment or a book imprint ---------
    tail_idx = None
    for i, seg in enumerate(body):
        if LOCATOR_RE.search(seg):
            tail_idx = i
            break
        bm = BOOK_TAIL_RE.match(seg)
        if bm:
            ref.is_book = True
            ref.place = bm.group("place").strip()
            ref.publisher = bm.group("publisher").strip()
            ref.year = ref.year or int(bm.group("year"))
            if bm.group("fp"):
                ref.first_page = ref.first_page or bm.group("fp")
                ref.last_page = ref.last_page or expand_page(bm.group("fp"), bm.group("lp"))
            tail_idx = i
            break

    if tail_idx is None:
        # No locator and no imprint: fall back to the last segment that is just
        # a year (e.g. "Lancet. 2024.") — then journal is the segment before it.
        for i in range(len(body) - 1, -1, -1):
            if re.fullmatch(r"(?:19|20)\d{2}\.?", body[i].strip()):
                tail_idx = i
                ref.year = ref.year or int(body[i].strip().rstrip("."))
                break

    if tail_idx is None:
        # last resort: title is everything, year from anywhere in the string
        ref.title = " ".join(body).rstrip(".").strip()
    elif ref.is_book:
        ref.title = " ".join(body[:tail_idx]).rstrip(".").strip()
    else:
        # segment immediately before the tail is the journal, unless the tail
        # segment itself carries the journal ("Lancet 2017;390(1):1-2")
        jrn_idx = tail_idx - 1
        tail_seg = body[tail_idx]
        pre_locator = tail_seg[: LOCATOR_RE.search(tail_seg).start()].strip(" ,.;") if LOCATOR_RE.search(tail_seg) else ""
        if pre_locator and not re.fullmatch(r"(?:19|20)\d{2}", pre_locator):
            ref.journal = pre_locator
            ref.title = " ".join(body[:tail_idx]).rstrip(".").strip()
        elif jrn_idx >= 0:
            ref.journal = body[jrn_idx].rstrip(".").strip()
            ref.title = " ".join(body[:jrn_idx]).rstrip(".").strip()
        else:
            ref.title = " ".join(body[:tail_idx]).rstrip(".").strip()

    if ref.year is None:
        years = YEAR_RE.findall(text)
        if years:
            ref.year = int(re.findall(r"\b(?:19|20)\d{2}\b", text)[-1])

    # a book with no imprint but an edition marker is still a book
    if not ref.is_book and ref.publisher is None and ref.volume is None and ref.edition:
        ref.is_book = True

    ref.title = re.sub(r"\s+", " ", ref.title).strip(" .,;")
    ref.journal = re.sub(r"\s+", " ", ref.journal).strip(" .,;")
    return ref


def parse_all(refs: dict[int, str]) -> dict[int, ParsedRef]:
    return {n: parse_reference(n, txt) for n, txt in refs.items()}
